### PR TITLE
RavenDB-16880 / RavenDB-14189 Power BI integration with the usage of PostgreSQL protocol

### PR DESCRIPTION
**Pull request by arekpalinski**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-14189
https://issues.hibernatingrhinos.com/issue/RavenDB-16880

### Additional description

Integration with Power BI via the usage of PostgreSQL protocol.

### Type of change

- New feature

### How risky is the change?

- Low 

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- Yes. It will be added to 5.3 docs.

### Testing 

- Tests have been added that prove the fix is effective or that the feature works
- It has been verified by manual testing

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- UI work is included in the PR

### CODE REVIEW

File: src/Raven.Client/Constants.cs [hunk 1]
```diff
@@ -334,6 +334,11 @@ private Facet()
 
                     public const string AllResults = "@AllResults";
                 }
+
+                public class Fields
```
Comment (ppekrol) [inline src/Raven.Client/Constants.cs]: can we make it internal?

File: src/Raven.Client/Documents/Operations/Integrations/ConfigurePostgreSqlOperation.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+using System.Net.Http;
+using Raven.Client.Documents.Conventions;
+using Raven.Client.Http;
+using Raven.Client.Json;
+using Raven.Client.Json.Serialization;
+using Raven.Client.ServerWide.Operations.Integrations.PostgreSQL;
+using Raven.Client.Util;
+using Sparrow.Json;
+
+namespace Raven.Client.Documents.Operations.Integrations
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Operations/Integrations/ConfigurePostgreSqlOperation.cs]: Do we want to put it in `.PostgreSQL` namespace?

File: src/Raven.Client/Documents/Operations/Integrations/ConfigurePostgreSqlOperation.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+{
+    public class ConfigurePostgreSqlOperation : IMaintenanceOperation<ConfigurePostgreSqlOperationResult>
+    {
+        private readonly PostgreSqlConfiguration _configuration;
+
+        public ConfigurePostgreSqlOperation(PostgreSqlConfiguration configuration)
+        {
+            _configuration = configuration;
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Operations/Integrations/ConfigurePostgreSqlOperation.cs]: check for null

File: src/Raven.Server/Integrations/PostgreSQL/Handlers/PostgreSqlIntegrationHandler.cs [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+using Raven.Server.ServerWide.Context;
+using Raven.Server.Utils;
+using Sparrow.Json;
+using Sparrow.Json.Parsing;
+
+namespace Raven.Server.Integrations.PostgreSQL.Handlers
+{
+    public class PostgreSqlIntegrationHandler : DatabaseRequestHandler
+    {
+        [RavenAction("/databases/*/admin/integrations/postgresql/users", "GET", AuthorizationStatus.ValidUser, EndpointType.Read)]
```
Comment (ppekrol) [inline src/Raven.Server/Integrations/PostgreSQL/Handlers/PostgreSqlIntegrationHandler.cs]: just to be sure - is this appropriate permission?

File: src/Raven.Server/ServerWide/ServerStore.cs [hunk 5]
```diff
@@ -1858,6 +1860,21 @@ public Task<(long Index, object Result)> ModifyDocumentsCompression(TransactionO
...
+        public Task<(long Index, object Result)> ModifyPostgreSqlConfiguration(TransactionOperationContext context, string databaseName, BlittableJsonReaderObject configurationJson, string raftRequestId)
+        {
+            var config = JsonDeserializationCluster.PostgreSqlConfiguration(configurationJson);
+
+            return ModifyPostgreSqlConfiguration(context, databaseName, config, raftRequestId);
+        }
+
+        public Task<(long Index, object Result)> ModifyPostgreSqlConfiguration(TransactionOperationContext context, string databaseName, PostgreSqlConfiguration configuration, string raftRequestId)
+        {
+            // TODO arek LicenseManager.assertcanusepo
```
Comment (ppekrol) [inline src/Raven.Server/ServerWide/ServerStore.cs]: TODO

File: src/Raven.Client/Documents/Smuggler/DatabaseItemType.cs [hunk 6]
```diff
@@ -58,6 +58,7 @@ public enum DatabaseRecordItemType
         OlapConnectionStrings = 1 << 19,
         OlapEtls = 1 << 20,
         ElasticSearchConnectionStrings = 1 << 21,
-        ElasticSearchEtls = 1 << 22
+        ElasticSearchEtls = 1 << 22,
+        PostgreSQLIntegration = 1 << 23,
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Smuggler/DatabaseItemType.cs:62]: IIRC in SmugglerHandler there is a method that adds missing items if you are connecting from previous Client API, we need to add that there

File: src/Raven.Client/Documents/Smuggler/DatabaseSmugglerOptions.cs [hunk 7]
```diff
@@ -36,7 +36,8 @@ public class DatabaseSmugglerOptions : IDatabaseSmugglerOptions
                                                                                   DatabaseRecordItemType.OlapConnectionStrings |
                                                                                   DatabaseRecordItemType.OlapEtls |
                                                                                   DatabaseRecordItemType.ElasticSearchConnectionStrings |
-                                                                                  DatabaseRecordItemType.ElasticSearchEtls;
+                                                                                  DatabaseRecordItemType.ElasticSearchEtls |
+                                                                                  DatabaseRecordItemType.PostgreSQLIntegration;
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Smuggler/DatabaseSmugglerOptions.cs:40]: IIRC in SmugglerHandler there is a method that adds missing items if you are connecting from previous Client API, we need to add that there

File: src/Raven.Client/Constants.cs (hunk 1 above)
Comment (arekpalinski) [reply src/Raven.Client/Constants.cs]: done

File: src/Raven.Client/Documents/Operations/Integrations/ConfigurePostgreSqlOperation.cs (hunk 2 above)
Comment (arekpalinski) [reply src/Raven.Client/Documents/Operations/Integrations/ConfigurePostgreSqlOperation.cs]: Yes. Moved

File: src/Raven.Client/Documents/Operations/Integrations/ConfigurePostgreSqlOperation.cs (hunk 3 above)
Comment (arekpalinski) [reply src/Raven.Client/Documents/Operations/Integrations/ConfigurePostgreSqlOperation.cs]: done

File: src/Raven.Server/Integrations/PostgreSQL/Handlers/PostgreSqlIntegrationHandler.cs (hunk 4 above)
Comment (arekpalinski) [reply src/Raven.Server/Integrations/PostgreSQL/Handlers/PostgreSqlIntegrationHandler.cs]: no, it wasn't. it got changed to:

```
[RavenAction("/databases/*/admin/integrations/postgresql/users", "GET", AuthorizationStatus.DatabaseAdmin)]
```

File: src/Raven.Server/ServerWide/ServerStore.cs (hunk 5 above)
Comment (arekpalinski) [reply src/Raven.Server/ServerWide/ServerStore.cs]: Done. Added the license assertions to the handler

File: src/Raven.Client/Documents/Smuggler/DatabaseItemType.cs (hunk 6 above)
Comment (arekpalinski) [reply src/Raven.Client/Documents/Smuggler/DatabaseItemType.cs:62]: I see `ApplyBackwardCompatibility()` method there. But we add only default db record types only if we have `DatabaseRecordItemType.None`:

https://github.com/ravendb/ravendb/blob/01a3c36aee7715da601ee33d77104cbee9c9186c/src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs#L159-L163

I don't see anything that would add it per `DatabaseRecordItemType` item

File: src/Raven.Server/Integrations/PostgreSQL/CsvToPg.cs [hunk 8]
```diff
@@ -0,0 +1,75 @@
+﻿using System;
+using System.Collections.Generic;
+using Microsoft.VisualBasic.FileIO;
+using Raven.Server.Integrations.PostgreSQL.Messages;
+
+namespace Raven.Server.Integrations.PostgreSQL
+{
+    public static class CsvToPg
```
Comment (ayende) [inline src/Raven.Server/Integrations/PostgreSQL/CsvToPg.cs:8]: What is this for?

File: src/Raven.Server/Integrations/PostgreSQL/Handlers/PostgreSqlUsernames.cs [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+    {
+        public List<User> Users { get; set; }
+
+        public PostgreSqlUsernames()
+        {
+            Users = new List<User>();
+        }
+    }
+
+    public class User
```
Comment (ayende) [inline src/Raven.Server/Integrations/PostgreSQL/Handlers/PostgreSqlUsernames.cs:15]: This is a *very* generic name. Can we rename that to something that won't likely collide with other things?

File: src/Raven.Server/Integrations/PostgreSQL/Npgsql/version_query.csv [hunk 10]
```diff
@@ -0,0 +1,2 @@
+"version"
+"PostgreSQL 13.3, compiled by Visual C++ build 1914, 64-bit"
```
Comment (ayende) [inline src/Raven.Server/Integrations/PostgreSQL/Npgsql/version_query.csv:2]: Are we providing this as an answer?

File: src/Raven.Server/Integrations/PostgreSQL/PgErrorCodes.cs [hunk 11]
```diff
@@ -0,0 +1,483 @@
+﻿/*
```
Comment (ayende) [inline src/Raven.Server/Integrations/PostgreSQL/PgErrorCodes.cs:1]: Provide a link for the source of this file.

File: src/Raven.Server/Integrations/PostgreSQL/PgServer.cs [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+                    activate = true;
+                }
+                else if (_server.ServerStore.LicenseManager.CanUsePostgreSqlIntegration(withNotification: true))
+                {
+                    if (_server.ServerStore.Configuration.Core.FeaturesAvailability == FeaturesAvailability.Experimental)
+                        activate = true;
+                    else
+                    {
+                        if (_logger.IsOperationsEnabled)
+                            _logger.Info($"You have enabled the PostgreSQL integration via '{RavenConfiguration.GetKey(x => x.Integrations.PostgreSql.Enabled)}' configuration but " +
```
Comment (ayende) [inline src/Raven.Server/Integrations/PostgreSQL/PgServer.cs:69]: Operations here, not info

File: src/Raven.Server/Integrations/PostgreSQL/PgSession.cs [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+
+            var messageReader = new MessageReader();
+
+            var initialMessage = await messageReader.ReadInitialMessage(reader, _token);
+
+            if (initialMessage is SSLRequest)
+            {
+                if (_serverCertificate == null)
+                {
+                    await writer.WriteAsync(messageBuilder.SSLResponse(false), _token);
```
Comment (ayende) [inline src/Raven.Server/Integrations/PostgreSQL/PgSession.cs:61]: What is the behavior in this case? We just reject the SSL connection?

File: src/Raven.Server/Integrations/PostgreSQL/PgSession.cs [hunk 14]
```diff
@@ -0,0 +1,232 @@
...
+                await writer.WriteAsync(messageBuilder.ErrorResponse(
+                    PgSeverity.Fatal,
+                    e.ErrorCode,
+                    e.Message,
+                    e.ToString()), _token);
+            }
+            catch (PgErrorException e)
+            {
+                if (Logger.IsInfoEnabled)
+                    Logger.Info($"{e.Message} (pg error code {e.ErrorCode})", e);
```
Comment (ayende) [inline src/Raven.Server/Integrations/PostgreSQL/PgSession.cs:200]: Provide details on the source connection?

File: src/Raven.Server/Integrations/PostgreSQL/PgSession.cs [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+
+                // Shouldn't get to this point, PgErrorExceptions shouldn't be fatal
+                await writer.WriteAsync(messageBuilder.ErrorResponse(
```
Comment (ayende) [inline src/Raven.Server/Integrations/PostgreSQL/PgSession.cs:203]: what happens when this throws?

File: src/Raven.Server/Integrations/PostgreSQL/CsvToPg.cs (hunk 8 above)
Comment (arekpalinski) [reply src/Raven.Server/Integrations/PostgreSQL/CsvToPg.cs:8]: This is used to handle the "hardcoded queries" about pg types, chracters set, table schema etc. The responses are generated based on results exported as CSV from PgAdmin as described here:

https://issues.hibernatingrhinos.com/issue/RavenDB-16880#focus=Comments-67-91291.0-0

File: src/Raven.Server/Integrations/PostgreSQL/Handlers/PostgreSqlUsernames.cs (hunk 9 above)
Comment (arekpalinski) [reply src/Raven.Server/Integrations/PostgreSQL/Handlers/PostgreSqlUsernames.cs:15]: Renamed to `PostgreSqlUsername` and pushed to already open PR:

https://github.com/ravendb/ravendb/pull/12989/commits/07e01aae221622dc916581adc3e5029eb0c03713

File: src/Raven.Server/Integrations/PostgreSQL/Npgsql/version_query.csv (hunk 10 above)
Comment (arekpalinski) [reply src/Raven.Server/Integrations/PostgreSQL/Npgsql/version_query.csv:2]: Yes, it's used as version response: 

https://github.com/ravendb/ravendb/blob/98f362180fb527c000ea90f958cd633b0fb7cb0e/src/Raven.Server/Integrations/PostgreSQL/Npgsql/NpgsqlConfig.cs#L14-L19

File: src/Raven.Server/Integrations/PostgreSQL/PgServer.cs (hunk 12 above)
Comment (arekpalinski) [reply src/Raven.Server/Integrations/PostgreSQL/PgServer.cs:69]: Fixed in https://github.com/ravendb/ravendb/pull/12989/commits/316d280ae7e484305a19d3051db1ecc64cd68a48

File: src/Raven.Server/Integrations/PostgreSQL/PgSession.cs (hunk 13 above)
Comment (arekpalinski) [reply src/Raven.Server/Integrations/PostgreSQL/PgSession.cs:61]: We say that we don't accept SSL so according to the docs - https://www.postgresql.org/docs/13/protocol-flow.html#id-1.10.5.7.11:

- "The frontend might close the connection at this point if it is dissatisfied with the response." 
- "To continue after N, send the usual StartupMessage and proceed without encryption. "

File: src/Raven.Server/Integrations/PostgreSQL/PgSession.cs (hunk 15 above)
Comment (arekpalinski) [reply src/Raven.Server/Integrations/PostgreSQL/PgSession.cs:203]: We log the error and close the tcp client for a given connection

File: src/Raven.Server/Integrations/PostgreSQL/PgSession.cs (hunk 14 above)
Comment (arekpalinski) [reply src/Raven.Server/Integrations/PostgreSQL/PgSession.cs:200]: What kind of details would to expect to see here?

Comment (ayende) [reply src/Raven.Server/Integrations/PostgreSQL/PgSession.cs:200]: IP, if certificate is used, its details?

Comment (ayende) [reply src/Raven.Server/Integrations/PostgreSQL/PgSession.cs:200]: Sorry, this is PGSQL - so IP & Username, I think

Comment (arekpalinski) [reply src/Raven.Server/Integrations/PostgreSQL/PgSession.cs:200]: Fixed in https://github.com/ravendb/ravendb/pull/12989/commits/4e7574a25f9bc3cad16382c3feebce0eb83ab916

**Pull request merged**